Replace dependsOn with @lifetime

meg-gupta opened the pull request:
- Adds support for specifying the target for lifetime dependence with : `@lifetime(target: source1...source)` 

- Delete `dependsOn` syntax

- Update all tests with `dependsOn` with `@lifetime`

meg-gupta committed:
Delete dependsOn support

Lifetime dependencies will now be represented with @lifetime attribute in the language.

dependsOn is a type modifier and was represented as a LifetimeDependentTypeRepr in the AST.

I am deleting dependsOn syntax parsing support and retaining LifetimeDependentTypeRepr support.

We may want to represent lifetime dependencies in a function type with a type attribute in the future.
If we use a decl attribute instead, then support for LifetimeDependentTypeRepr can be deleted.

meg-gupta committed:
Fix AST printing of @lifetime

meg-gupta committed:
Update some dependsOn tests to @lifetime

meg-gupta committed:
Add support for specifying target in @lifetime

@lifetime(target: source1, source2...) where target can be any
parameter or 'self'. We cannot have @lifetime attributes with duplicate targets.

Also, update the internal data structures. Previously LifetimeEntry stored
pairwise (target, source) dependencies. Now, LifetimeEntry will store an optional
target descriptor and an array of source descriptors.

meg-gupta committed:
Update SIL printing of @lifetime

Lifetime dependencies in SIL tests continue to be represented as a type modifier on the target.

As before, they are represented as a LifetimeDependentTypeRepr in the AST.

meg-gupta committed:
Diagnose multiple @lifetime attributes with same target

meg-gupta committed:
Add a test for @lifetime(immortal)

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta committed:
Fix SIL parsing of lifetime dependence in parameter position

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

ahoppen commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -5179,109 +5136,80 @@ ParserStatus Parser::parseTypeAttribute(TypeOrCustomAttr &result,
+  auto lParenLoc = consumeToken();
+
+  std::optional<LifetimeDescriptor> targetDescriptor;
+  if (Tok.isAny(tok::identifier, tok::integer_literal, tok::kw_self) &&
```
I just want to clarify. Is `@lifetime(self: something)` and `@lifetime(1: something)` considered valid syntax? Because the latter won’t parse in the new parser since `1` is not a valid argument label.

meg-gupta replied:
We were planning to use integer indices in the case of `dependsOn` especially for function types which may not have parameter names. `@lifetime(1: 0)` doesn't read as well as a positional `dependsOn(0)` would.  We likely can delete support for integer indices. I'll follow with a separate PR for this.

ahoppen replied:
I would be in favor of deleting integer target indices.

ahoppen commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -5179,109 +5136,80 @@ ParserStatus Parser::parseTypeAttribute(TypeOrCustomAttr &result,
 
-    if (!Tok.isFollowingLParen()) {
-      diagnose(Tok, diag::expected_lparen_after_lifetime_dependence);
+  if (!Tok.isFollowingLParen()) {
```
Could you disallow spaces between the attribute name and the `(`, similar to what I did in https://github.com/swiftlang/swift/pull/71237?

meg-gupta replied:
Ok, I am now using `consumeAttributeLParen()`

atrick reviewed (approved):
This all looks good to me, except the one comment about BitwiseCopyable. I barely skimmed the syntax support since I'm not familiar with that. It looks like @ahoppen reviewed that part, thanks!

meg-gupta committed:
[NFC] Use llvm::TrailingObjects to represent variable list of lifetime sources

meg-gupta committed:
Don't allow space between @lifetime and '('

meg-gupta committed:
Disallow target specification in SIL's @lifetime

SIL's @lifetime continues to be positional. Don't parse 'target:' in this case.

meg-gupta commented:
@ahoppen https://github.com/swiftlang/swift-syntax/pull/2876 deletes `dependsOn` support in swift-syntax

meg-gupta commented:
@swift-ci test
